src/core/project/qgsproject.cpp

```diff
@@ -1563,6 +1597,53 @@ void QgsProject::releaseHandlesToProjectArchive()
   mStyleSettings->removeProjectStyle();
 }
 
+bool QgsProject::rebuildCrs3D( QString *error )
+{
+  bool res = true;
+  if ( !mCrs.isValid() )
+  {
+    mCrs3D = QgsCoordinateReferenceSystem();
+  }
+  else if ( !mVerticalCrs.isValid() )
+  {
+    mCrs3D = mCrs;
+  }
+  else
+  {
+    switch ( mCrs.type() )
+    {
+      case Qgis::CrsType::Compound:
+      case Qgis::CrsType::Geographic3d:
+      case Qgis::CrsType::Geocentric:
+        mCrs3D = mCrs;
+        break;
+
+      case Qgis::CrsType::Vertical:
+        // nonsense situation
+        mCrs3D = QgsCoordinateReferenceSystem();
+        res = false;
+        break;
+
+      case Qgis::CrsType::Unknown:
+      case Qgis::CrsType::Geodetic:
+      case Qgis::CrsType::Geographic2d:
+      case Qgis::CrsType::Projected:
```

Comment:
some Projected CRS might already be 3D. You should probably check the axis count

Reply:
@rouault 
Thanks!

Does [458b1fd](https://github.com/qgis/QGIS/pull/57107/commits/458b1fde397cabd1f268c58ad248836c0b602e5c) and https://github.com/qgis/QGIS/pull/57107/commits/6773e679c1d195cc0aeb9a4ada75fc53449af853 address this correctly?

Reply:
yes, looks good

Reply:
Thanks @rouault !